### PR TITLE
New style copy function

### PR DESCRIPTION
**Pull request by okuta**
Please see #3147.

### CODE REVIEW

File: chainer/functions/array/copy.py [hunk 1]
```diff
@@ -15,41 +15,17 @@ def check_type_forward(self, in_types):
             in_types.size() == 1
         )
 
-    def forward_cpu(self, x):
-        self.retain_inputs(())
-        self._in_device = cuda.get_device_from_array(x[0])
+    def forward(self, inputs):
+        x, = inputs
+        self._in_device = cuda.get_device_from_array(x).id
+        print(type(x), self._in_device, self.out_device)
```
Comment (niboshi) [inline chainer/functions/array/copy.py]: Remove `print`.

File: chainer/functions/array/copy.py [hunk 2]
```diff
@@ -83,4 +59,6 @@ def copy(x, dst):
         -1
 
     """
-    return Copy(dst)(x)
+    assert isinstance(dst, int)
```
Comment (niboshi) [inline chainer/functions/array/copy.py]: Shouldn't it accept `Device` as `dst`?

Comment (okuta) [reply chainer/functions/array/copy.py]: Maybe, only accept `int`.
https://github.com/chainer/chainer/pull/3419/files#diff-1b7bc2d145846eadbc5d0f8a8fd90c20R22

Comment (niboshi) [reply chainer/functions/array/copy.py]: Is there any reason?
IMO, it would be more convenient for user if it accepts `Device` as well.

Comment (okuta) [reply chainer/functions/array/copy.py]: This is my mistake.
I added test. and I removed this assert.

**Comment by okuta** [comment]
I fixed.

**Comment by niboshi** [comment]
Please add double-backward tests.

**Comment by niboshi** [comment]
LGTM!

**Pull request merged**